packages/core/core/src/BundleGraph.js

```diff
@@ -253,37 +336,69 @@ export default class BundleGraph {
     );
   }
 
-  isAssetReferencedByAnotherBundleOfType(asset: Asset, type: string): boolean {
-    let referringBundles = new Set(
-      this._graph.getNodesConnectedTo(
-        nullthrows(this._graph.getNode(asset.id)),
-        'contains',
-      ),
-    );
-
-    // is `asset` referenced by a dependency from an asset of `type`
-    return flatMap(
-      this._graph
-        .getNodesConnectedTo(nullthrows(this._graph.getNode(asset.id)))
-        .filter(node => {
-          // Does this dependency belong to a bundle that does not include the
-          // asset it resolves to? If so, this asset is needed by a bundle but
-          // does not belong to it.
-          return this._graph
-            .getNodesConnectedTo(node, 'contains')
-            .filter(node => node.type === 'bundle')
-            .some(b => !referringBundles.has(b));
-        }),
-      node => {
+  isAssetReferencedByDependant(
+    bundle: Bundle,
+    asset: Asset,
+    visitedBundles: Set<Bundle> = new Set(),
+  ): boolean {
+    let dependencies = this._graph
+      .getNodesConnectedTo(nullthrows(this._graph.getNode(asset.id)))
+      .filter(node => node.type === 'dependency')
+      .map(node => {
         invariant(node.type === 'dependency');
-        return this._graph.getNodesConnectedTo(node, null);
-      },
-    )
-      .filter(node => node.type === 'asset')
-      .some(node => {
-        invariant(node.type === 'asset');
-        return node.value.type === type;
+        return node.value;
       });
+
+    const bundleHasReference = (bundle: Bundle) => {
+      return (
+        !this.bundleHasAsset(bundle, asset) &&
+        dependencies.some(dependency =>
+          this.bundleHasDependency(bundle, dependency),
+        )
+      );
+    };
+
+    let isReferenced = false;
+    this.traverseBundles((descendant, _, actions) => {
+      if (visitedBundles.has(descendant)) {
+        actions.skipChildren();
+        return;
+      }
+
+      visitedBundles.add(descendant);
+      if (
+        descendant.type !== bundle.type ||
+        descendant.env.context !== bundle.env.context
+      ) {
+        actions.skipChildren();
+        return;
+      }
+
+      if (descendant !== bundle && bundleHasReference(descendant)) {
+        isReferenced = true;
+        actions.stop();
+        return;
+      }
+
+      let similarSiblings = this.getSiblingBundles(descendant).filter(
```

Comment:
Does this fix the first issue in #4295?

Reply:
I think that issue is an issue in `isAssetReferencedByAncestorBundle` — the inverse of this function — where we don't detect at deduplication time that the asset _can_ be referenced, so we leave it in the bundle.